### PR TITLE
resolve "BUG: existing user joining group with oauth broken (#657)"

### LINKED ISSUE #657: BUG: on joining a group (via google or fbook) and using a phone number that already exists in the DB, it returns fields that were not previously there

### PR DESCRIPTION
**Pull request by aguestuser**
resolves #657 
________________

# Functionality

Handle OAuth group-joining flow corner cases by checking in "step 1" to see if a person with email matching OAuth email already exists, then:

* if person does not yet exist, handle as before
* if they do exist:
  * if they are already member of group: log them in and redirect them to group dashboard
  * if they aren't member and lack contact info: redirect to members/edit` form where we gather their contact info, make them a member, and record their oauth info (but not until then)
  * if they aren't member and have all contact info: don't bother redirecting. just make them a member, record their token, and log them into group dashboard page

Handle email group-joining flow corner cases as follows:

* if user follows email join path and submits form with already-existing email, redirect them to login page with error message
* after they login, redirect them back to group join page, where they may join with one click

# Manual Testing Notes
(see forthcoming comment. it's hard to create the state necessary to test all these corner cases)

### CODE REVIEW

**Comment by aguestuser** [comment]
note: this breaks some tests for existing `edit member` functionality. marking as blocked and returning to clean it up tomorrow. :/

**Comment by aguestuser** [comment]
fixed broken `members/edit` functionality and did some fancy git footwork (cherry-picking, rebasing, etc.) to make sure all commits are in working state.

**Comment by aguestuser** [comment]
# Notes for Manual Testing

To verify this works for all corner cases, do the following:

## Setup

* Login to both Facebook and Gmail with testy@affinity.works
* In the console, use the class method `Person.find_by_email("testy@affinity.works")` to ensure that no user with that email already exists

## Testing New User Path

For Google:
* go to `/groups/1/join`
* click "Join with Google"
* verify that you are logged in and forwarded to dashboard after entering required info
* log out
* in console, use `Person.find_by_email("testy@affinity.works").destroy` to destroy testy

For Facebook:
* go to `/groups/1/join`
* click "Join with Facebook"
* verify that you are logged in and forwarded to dashboard after entering required info
* log out
* in console, use `Person.find_by_email("testy@affinity.works").destroy` to destroy testy

For Email:
* go to `/groups/1/join`
* click "Join with Email"
* verify that you are logged in and forwarded to dashboard after entering required info
* log out
* in console, use `Person.find_by_email("testy@affinity.works").destroy` to destroy testy

## Testing Existing User Paths



##### For Email

* SETUP: ensure that a user with `testy@affinity.works` already exists in the system, but does *not* yet belong to any groups
* go to `/groups/1/join`
* click "Join with Email"
* fill out form
* verify that you are redirected to login page
* login with testy@affinity.works
* verify that you are redirected to `groups/1/join` after logging in
* verify that you see a "join" button
* verify that after clicking "join" you are redirected to group dashboard

### For Facebook/Google

#### When user already belongs to group

* SETUP: piggy-pack off of prior step or add testy to National Network with:
```
testy = Person.find_by_email("testy@affinity.works")
Membership.create(group: Group.first, person: testy)
```
* go to `/groups/1/join`
* click "Join with Facebook" / "Join with Google"
* verify that you are logged in and redirected to dashboard
* logout
* repeat for other OAuth method

#### When user does not belong to group

##### And is missing no contact info

* SETUP: in console, ensure that `testy@affinity.works` user exists, is not member of `Group.first`, and has a zip code
* go to `groups/1/join`
* click "Join with Google" / "Join with Facebook"
* verify that you are logged in and redirected to the group dashboard page
* repeat for other OAuth method

##### And is missing contact info

* SETUP: in console, ensure that `testy@affinity.works` user exists, is not member of `Group.first`, and lacks a zipcode. (you can delete a zipcode with person.primary_personal_address.update!(postal_code: nil)`
* go to `groups/1/join`
* click "Join with Google" / "Join with Facebook"
* verify that you are redirected to the `members/edit` form
* fill in missing zipcode and click "submit"
* verify that you are logged in and redirected to group dashboard page
* repeat for other OAuth method

**Comment by Matt-Holland** [comment]
# Manual Testing Results:

Testing at: https://staging-affinityworks-pr-669.herokuapp.com/groups/2/join

*Fail:* at step I.A.2.  result is: 
`Error: redirect_uri_mismatch
The redirect URI in the request, https://staging-affinityworks-pr-669.herokuapp.com/admin/auth/google_oauth2/callback, does not match the ones authorized for the OAuth client.`

*Fail:* at step I.B.2. result is:
`URL Blocked: This redirect failed because the redirect URI is not whitelisted in the app’s Client OAuth Settings. Make sure Client and Web OAuth Login are on and add all your app domains as Valid OAuth Redirect URIs.`

*NOTE:* Suspended further testing until above issues can be addressed.

**Comment by aguestuser** [comment]
@Matt-Holland : i added redirect URIs for the review app to both FB and Google OAuth admin panels and updated the docs to make it clearer how to do that and what steps are not needed to do so. should be good to go.

File: app/controllers/people/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -1,18 +1,23 @@
 class People::OmniauthCallbacksController < Devise::OmniauthCallbacksController
   before_action :set_auth
   before_action :set_auth_mode
-  before_action :set_group_id
 
   def facebook
-    auth "facebook"
+    @service = "facebook"
+    send @auth_mode
```
Comment (vaughanj10) [inline app/controllers/people/omniauth_callbacks_controller.rb]: I find using send for dynamic method calling is pretty confusing here as it is non obvious what the possible auth_modes are here, and even what auth_mode is.

File: app/controllers/people/omniauth_callbacks_controller.rb [hunk 2]
```diff
@@ -1,18 +1,23 @@
...
   end
 
   def google_oauth2
-    auth "google"
+    @service = "google"
+    send @auth_mode
```
Comment (vaughanj10) [inline app/controllers/people/omniauth_callbacks_controller.rb]: Same as comment on line 7.

File: app/models/person.rb [hunk 3]
```diff
@@ -289,6 +299,16 @@ def import_remote(remote_people, group, event, member_id)
     end # import_remote
   end # CLASS METHODS
 
+  #
+  # MUTATIONS
+  #
+
+  def update_from_oauth_signup(auth, person_params={})
+    self.class.save_omniauth(EmailAddress.find_by(address: auth.info.email),
```
Comment (vaughanj10) [inline app/models/person.rb]: I feel like it would be more clear to just use `Person.save_omniauth` here

File: app/models/person.rb [hunk 4]
```diff
@@ -212,22 +222,22 @@ def find_first_by_auth_conditions(warden_conditions)
...
       email, given, family = parse_info_from_auth(auth)
       return unless email
-      if person = Person.find_by_email(email)
-        save_omniauth(EmailAddress.find_by(address: email),
-                      Identity.find_for_oauth(auth),
-                      person.tap{ |p| p.update(person_params)})
-      else
-        create new_signup_attrs(auth, person_params, email, given, family)
-      end
+      send constructor, signup_attrs(auth, person_params, email, given, family)
```
Comment (vaughanj10) [inline app/models/person.rb]: I feel that using `send` here obscures the meaning of what is trying to be done more than it helps saving the two lines of code needed to added to `build_from_oauth_signup` and `create_from_oauth_signup`

File: app/models/person.rb (hunk 3 above)
Comment (aguestuser) [reply app/models/person.rb]: okay. i prefer to not use `Person`. IMHO `self.class` has very clear semantics. what do we do?

File: app/controllers/members_controller.rb [hunk 5]
```diff
@@ -132,7 +143,7 @@ def person_params
               credentials: [:token, :expires_at, :expires ],
               info:        [:email, :name, :image ] ]
     ).tap{ |person_attrs| handle_empty_contact_info(person_attrs) }
-      .tap { |person_attrs| maybe_set_role(person_attrs) }
+      .tap { |person_attrs| maybe_set_memberships(person_attrs) }
```
[the remark this answers is not in the document]
Comment (aguestuser) [reply app/controllers/members_controller.rb]: sure. like `tap { |person_attrs| format_person_attrs }` which dispatches to `maybe_set_role` and `maybe_set_memberships`?

only thing i think is important is to keep the two helpers distinct since they do two different things.

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -9,8 +9,20 @@ def current_user
   end
 
   def current_group
-    group_id = controller_name == 'groups' ? params[:id] : params[:group_id]
-    Group.find(group_id) if group_id
+    if group_id = parse_group_id
```
[the remark this answers is not in the document]
Comment (aguestuser) [reply app/controllers/application_controller.rb]: i disagree. i *really* like the pattern of setting inside conditionals, which is a common pattern in ruby i've read, go, and many other languages. why don't you prefer it?

File: app/models/person.rb (hunk 4 above)
Comment (aguestuser) [reply app/models/person.rb]: i disagree, but less strongly than in other cases. perhaps we could pass `Person.new` and `Person.create` as lambdas and then `call` them here?

i don't think it's worth repeating ourself in this logic to avoid "digging" two lines up into a caller method that very clearly passes either `new` or `create` to this helper... also, i changed this logic several times in two places, which is what led me to extract it into one place.

**Comment by aguestuser** [comment]
@vaughanj10 : made changes we discussed, left some suggestions unfielded, but open to revisiting as noted in comments. :D

good to :ship: ?

File: app/models/person.rb (hunk 3 above)
Comment (aguestuser) [reply app/models/person.rb]: :heavy_check_mark:

File: app/controllers/people/omniauth_callbacks_controller.rb (hunk 2 above)
Comment (aguestuser) [reply app/controllers/people/omniauth_callbacks_controller.rb]: revised.

File: app/controllers/people/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (aguestuser) [reply app/controllers/people/omniauth_callbacks_controller.rb]: but i changed my mind! revised. :)

File: app/controllers/members_controller.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (aguestuser) [reply app/controllers/members_controller.rb]: whoops. forgot to field this one. think it's okay to leave on the table for now. if we wind up with a bunch of `tap`s piling up (ie: one more than this), let's group them into one.

File: app/controllers/application_controller.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (aguestuser) [reply app/controllers/application_controller.rb]: agree to disagree for now. completely open to having my mind changed.

**Comment by Matt-Holland** [comment]
**Fail** in step II.4.iv.  -- I'm redirected to the /home page which lists all the groups, instead of to the dashboard of the group I just joined.

**Comment by Matt-Holland** [comment]
all the other paths are tested and work 💯
